lib/filterx/func-unset-empties.c

```diff
@@ -216,24 +280,154 @@ _extract_object_expr(FilterXFunctionArgs *args, GError **error)
 
   if (eval_error)
     {
       g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
-                  "recursive argument must be boolean literal. " FILTERX_FUNC_UNSET_EMPTIES_USAGE);
+                  FILTERX_FUNC_UNSET_EMPTIES_ARG_NAME_IGNORECASE" argument must be boolean literal. " FILTERX_FUNC_UNSET_EMPTIES_USAGE);
       return FALSE;
     }
 
-  self->recursive = value;
+  set_flag(&self->flags, FILTERX_FUNC_UNSET_EMPTIES_FLAG_IGNORECASE, value);
+
   return TRUE;
 }
 
+
+static gboolean
+_extract_optional_arg_replacement(FilterXFunctionUnsetEmpties *self, FilterXFunctionArgs *args, GError **error)
+{
+  gboolean exists;
+  FilterXObject *value = filterx_function_args_get_named_literal_object(args,
+                         FILTERX_FUNC_UNSET_EMPTIES_ARG_NAME_REPLACEMENT, &exists);
+  if (!exists)
+    return TRUE;
+  if (!value)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  FILTERX_FUNC_UNSET_EMPTIES_ARG_NAME_REPLACEMENT" argument must be literal. " FILTERX_FUNC_UNSET_EMPTIES_USAGE);
+      return FALSE;
+    }
+
+  self->replacement = value;
+  return TRUE;
+}
+
+static gboolean
+_handle_target_object(FilterXFunctionUnsetEmpties *self, FilterXObject *target, GError **error)
+{
+  g_assert(target);
+  guint64 len;
+  if (filterx_object_is_type(target, &FILTERX_TYPE_NAME(null)))
+    set_flag(&self->flags, FILTERX_FUNC_UNSET_EMPTIES_FLAG_REPLACE_NULL, TRUE);
+  else if (filterx_object_is_type(target, &FILTERX_TYPE_NAME(list)))
+    {
+      g_assert(filterx_object_len(target, &len));
+      if (len == 0)
+        set_flag(&self->flags, FILTERX_FUNC_UNSET_EMPTIES_FLAG_REPLACE_EMPTY_LIST, TRUE);
+    }
+  else if (filterx_object_is_type(target, &FILTERX_TYPE_NAME(dict)))
+    {
+      g_assert(filterx_object_len(target, &len));
+      if (len == 0)
+        set_flag(&self->flags, FILTERX_FUNC_UNSET_EMPTIES_FLAG_REPLACE_EMPTY_DICT, TRUE);
+    }
+  else if (filterx_object_is_type(target, &FILTERX_TYPE_NAME(string)))
+    {
+      const gchar *str = filterx_string_get_value(target, &len);
+      if (len == 0)
+        set_flag(&self->flags, FILTERX_FUNC_UNSET_EMPTIES_FLAG_REPLACE_EMPTY_STRING, TRUE);
+      else
+        {
+          if (check_flag(self->flags, FILTERX_FUNC_UNSET_EMPTIES_FLAG_IGNORECASE))
+            g_ptr_array_add(self->targets, g_utf8_casefold(str, len));
+          else
+            g_ptr_array_add(self->targets, g_strdup(str));
+        }
+    }
+  else
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  FILTERX_FUNC_UNSET_EMPTIES_ARG_NAME_TARGETS" has unknown target type: %s", target->type->name);
+      return FALSE;
+    }
+  return TRUE;
+}
+
+static gboolean
+_target_iterator(FilterXExpr *key, FilterXExpr *value, gpointer user_data)
+{
+  FilterXFunctionUnsetEmpties *self = ((gpointer *) user_data)[0];
+  GError **error = ((gpointer *) user_data)[1];
+
```

Comment:
First let's check whether the expression itself is literal, because if it is a function or variable, we should fail.